### ISSUE TITLE
Adding Lag support

### ISSUE THREAD

**Issue by ninjapapa** [TresAmigosSD/SMV#128]
It is quite common to calculate differences between current record and the previous records for a given time serious. The most familiar approach for that requirement is the lag function in SAS. 
Using GDO/CDS concept we can support Lag easily. The question is how the client code should look like. Here is one possibility:
```scala
df.smvGroupBy("acct").lagSelect($"date")($"amt" - $"amt".lag)
```

Another thing need to be figured out is how to handle the first record.

**Comment by ninjapapa** [comment]
Actually existing CDS can do this:

```scala
df.smvGroupBy("k").runAgg("t")(
  first($"amt") from InLastN(2) as "amt_lag"
)
```

will give you the lag(1). To make it a little explicit, we can create a ColumnHelper method ```smvLag```, and the client code will look like:

```scala
    val res = srdd.smvGroupBy("k").runAgg("t")(
      $"k",
      $"t",
      $"v",
      $"v".smvLag(1) as "v_lag"
   )
``` 

With this implementation, for each group, the lag value of the very first record will be the same as the un lagged value. Here is an example output of above code:
```scala
[a,1,0.3,0.3]
[z,1,0.2,0.2]
[z,2,1.4,0.2]
[z,5,2.2,1.4]
```
where "a,1" and "z,1" record has "v" === "v".smvLag(1). 

Another alternative is to make "v".smvLag(1) of the first record "null". To do so, we could implement another version of "InLastN" CDS.

**Comment by ninjapapa** [comment]
Since filling the current value instead of null can't be undo, but filling in null could aways be default to current value on client code, we decided to change the ```smvLag``` behavior to return null for the first record. 

In the process of implementing this, we found that ```first``` aggregation function of Spark actually return the first NON-NULL value, which could be a bug. We implemented ```smvFirst``` to really perform the real  first operation, which return the first record value regardless of it's null or not. 

The client code is still the same. 

The only limitation is that we can apply expression operations within ```runAgg```. So to calculate the difference between current value and lagged value, user need to have a ```select``` or ```selectPlus``` step after ```runAgg``` step, like the following

```scala
val res = srdd.smvGroupBy("k").runAgg("t")(
      $"k",
      $"t",
      $"v",
      $"v".smvLag(1) as "v_lag"
   ).selectPlus($"v" - $"v_lag" as "v_increase")
```
